File: 백준/Gold/14267. 회사 문화 1/회사 문화 1.py
import sys
sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline

def dfs(start, w):
    dist[start] += w

    for next in tree[start]:
        dfs(next, dist[start])

N, M = map(int, input().split())

tree = [[] for _ in range(N + 1)]
dist = [0] * (N + 1)

arr = list(map(int, input().split()))

for i in range(1, len(arr)):
    tree[arr[i]].append(i + 1)

for _ in range(M):
    a, b = map(int, input().split())

    dist[a] += b

dfs(1, 0)

print(*dist[1:], sep=' ')

File: 백준/Gold/14267. 회사 문화 1/회사 문화 1.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;

public class Main {

    static int N, M;

    static List<ArrayList<Integer>> graph;
    static int[] dist;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] temp = br.readLine().split(" ");
        N = Integer.parseInt(temp[0]);
        M = Integer.parseInt(temp[1]);

        temp = br.readLine().split(" ");

        graph = new ArrayList<>();
        dist = new int[N + 1];

        for(int i = 0; i <= N; i++){
            graph.add(new ArrayList<>());
        }

        for(int i = 1; i < N; i++){
            int num = Integer.parseInt(temp[i]);
            graph.get(num).add(i + 1);
        }

        for(int i = 0; i < M; i++){
            temp = br.readLine().split(" ");

            int start = Integer.parseInt(temp[0]);
            int w = Integer.parseInt(temp[1]);

            dist[start] += w;
        }

        dfs(1,0);

        for(int i = 1; i <= N; i++){
            System.out.print(dist[i] + " ");
        }


    }

    public static void dfs(int start, int w){
        dist[start] += w;
        for(int next : graph.get(start)){
            dfs(next, dist[start]);
        }
    }
}
